src/cpu/aarch64/acl_gemm_convolution.hpp

```diff
@@ -144,10 +139,9 @@ struct acl_gemm_convolution_fwd_t : public primitive_t {
     }
 
 private:
-    // To guard the const execute_forward(), the mutex must be 'mutable'
-    mutable std::mutex mtx;
     status_t execute_forward(const exec_ctx_t &ctx) const;
     const pd_t *pd() const { return (const pd_t *)primitive_t::pd().get(); }
+    std::unique_ptr<acl_obj_t<arm_compute::NEGEMMConvolutionLayer>> acl_obj_;
```

Comment:
Hi @snadampal, as far as I remember the main reason of putting the ACL objects into resources was to avoid storing them in the `primitive_t` because those ACL objects are not immutable. All the entities that reside in `primitive_t` must be immutable due to primitive cache design, otherwise there might be problems with concurrent execution of identical primitives that were fetched from the primitive cache.

Per my understanding the ACL object is still not immutable therefore we have to keep it in the resource. Am I missing something?

Reply:
Hi @densamoilov , could you please elaborate on this? What do you mean by identical primitives here?

>  otherwise there might be problems with concurrent execution of identical primitives that were fetched from the primitive cache.

I have verified these changes with PyTorch. However, with TensorFlow I see out of order execution and mismatch for the src and weight tensor pair randomly. I'm suspecting it due to concurrent execution of the cached primitives, and debugging it right now.
Btw, these changes provide significant performance improvement for CNNs with large number of channels, hence I would like to better understand the concurrent execution constraint and explore options to handle them.

Reply:
> Per my understanding the ACL object is still not immutable therefore we have to keep it in the resource. Am I missing something?

Yep, this was the original motivation back when we first discussed ACL integration - nothing has changed on the ACL side since then.

Reply:
> I would like to better understand the concurrent execution constraint and explore options to handle them.

The primitive cache is a global object that stores created primitive implementations with `shared_ptr`. A primitive implementation is a class derived from `primitive_t`. There is a difference between primitive and primitive implementation. A primitive is a class named `primitive_iface_t` that includes the primitive implementation (`primitive_t`), scratchpad and resource. The `primitive_iface_t` is created every time when a user creates a primitive. In fact, a pointer to `primitive_iface_t` is what the library returns to the user. Since primitive implementations can be shared between threads we have to make sure that they are immutable to avoid race condition.

It can easily happen that a user creates and executes two identical primitives in different threads.
```cpp
#pragma omp parallel num_threads(2)
{
    // While the convolution handles are different they both point to the
    // same primitive implementation internally.
    convolution_forward conv({...});
    // If the ACL object is stored in the primitive implementation then this will
    // lead to a race condition.
    conv.execute(...);
}
```
In order to address such issues we introduced an abstraction called resource. The resource is a storage that is unique for each created primitive and is used to store data that cannot be shared.

Reply:
Thanks for the details,@densamoilov. Yes, I have seen this note for primitive iface class 
https://github.com/oneapi-src/oneDNN/blob/v2.6/src/common/primitive.hpp#L270
that's when I started moving the acl object from resource to primitive so that it will get cached.

And from the below note for the resource class, I didn't infer acl objects be part of this cache exception list.
https://github.com/oneapi-src/oneDNN/blob/v2.6/src/common/primitive.hpp#L189

thanks for the code snippet above, it makes things more clear. So, for identical primitives, 
1. there will be a single primitive implementation object but having multiple resources (and hence the multiple acl object) associated with it, right? 
2. they are differentiated through the resource mapper object set in each execution context, right?

I'm curious how it is handled for other engines (eg. gpu) where resource allocation will be an overhead every time. and I didn't see the "resource_t" implementation for cpu/x86 engine. How is it handled there?

Reply:
Just a note: it seems that the acl object have to reside in ressource. However, the allocated memory does not have to.
If you rely on scratchpad for memory allocation (how it should be, given how they are used), the memory allocation would no more happen upon each execution. At execution time, the pointers in scratchpad would be imported to the acl tensor objects. Does this make sense?

A pre-requisite though would be for acl tensors to have a method to query size (in order to know the size of scratchpad).

Reply:
Yes, looks like scratchpad or acl memory pool is the option I see to avoid redundant memory allocations. Thanks @mgouicem.

Reply:
> 1. there will be a single primitive implementation object but having multiple resources (and hence the multiple acl object) associated with it, right?

Yes. 
> 2. they are differentiated through the resource mapper object set in each execution context, right?

Not exactly. I simplified my explanation when I said there is 1 resource instance per primitive. In practice it's possible that a primitive implementation includes another primitive implementation and that primitive implementation includes another one and so on. We call them nested primitive implementations. In the case of ACL each of them may have their own ACL object and therefore their own resource instance. This requires a mechanism whereby the nested primitive implementations can obtain their resource instances. The mechanism is implemented via the resource mapper abstraction, where key is address of a primitive implementation and value is the corresponding resource instance.
To sum it up:
* There is always only 1 instance of resource mapper **per primitive**
* In the case when **there are no** nested primitive implementations the resource mapper contains only 1 resource instance for the main primitive implementation
* In the case when **there are** nested primitive implementations the resource mapper contains resource instance for the main primitive implementation and all nested ones

If we want to cache ACL objects then we either need to be able to clone them (similar to OpenCL kernels) at execution time or make sure that they are immutable. I don't see other ways to cache them.

> I didn't see the "resource_t" implementation for cpu/x86 engine. How is it handled there?

That's because for CPU we have JIT kernels, which are represented as immutable arrays of bytes therefore we can cache them.

Reply:
@densamoilov thanks! I'm exploring few options with ACL memory manager level caching the allocations outside the acl object. Will update this PR once the changes are ready for review.